### ISSUE TITLE
Error: [vuex] do not mutate vuex store state outside mutation handlers

### ISSUE THREAD

**Issue by mhluska** [mrichar1/jsonapi-vuex#26]
I'm trying to run this with [strict mode](https://vuex.vuejs.org/guide/strict.html) but notice the following error:

```js
vue.runtime.esm.js?2b0e:1887 Error: [vuex] do not mutate vuex store state outside mutation handlers.
    at assert (vuex.esm.js?2f62:87)
    at Vue.store._vm.$watch.deep (vuex.esm.js?2f62:763)
    at Watcher.run (vue.runtime.esm.js?2b0e:4529)
    at Watcher.update (vue.runtime.esm.js?2b0e:4503)
    at Dep.notify (vue.runtime.esm.js?2b0e:730)
    at Function.del [as delete] (vue.runtime.esm.js?2b0e:1125)
    at actionStatusClean (jsonapi-vuex.js?1c7f:360)
```

It looks like the `actionStatusClean` function is doing some direct state modifications?

**Comment by mrichar1** [comment]
Yes - this is an issue with the way that the cleanup works, which I've been racking my brains to find a solution for. I was hoping I'd have one before someone turned on `strict` mode :grin: 

Since the `setInterval`timer is fired off when the module is instantiated, it has no access to the `mutations` so can't use an e.g. `remove_status` mutation to update the store. Instead I'm just relying on using `Vue.set` to make sure Vue sees the deletion, but this prevents the use of `strict` which only allows modification via mutations.

I *think* the only solution to this is to not start the timer on cleanup instantiation, and instead expect the module user to call the method (either `actionStatusClean` directly, or an indirect method to start the timer), passing in a store context object so that the cleanup can use mutations. (Though I am still thinking about alternatives!)

`strict` mode is a devlopment-only tool for debugging, and the docs say:

> **Do not enable strict mode when deploying for production!** Strict mode runs a synchronous deep watcher on the state tree for detecting inappropriate mutations, and it can be quite expensive when you make large amount of mutations to the state. Make sure to turn it off in production to avoid the performance cost.

So this hasn't been a high priority for me to resolve, but it is on the radar, so thanks for flagging it up!

**Comment by mhluska** [comment]
Yeah, not a huge deal. I wish I had some better advice on how to solve this but I'm still relatively new to Vue. Two thoughts come to mind:

- Is it necessary to periodically clean these status records out in the first place? Is there a way to achieve status tracking without using those records?
- Would using a [plugin](https://vuex.vuejs.org/guide/plugins.html) instead of a module allow you to access mutations sooner?

**Comment by mrichar1** [comment]
Yes, it's purely to prevent memory 'leaks'. You can avoid it easily by setting `action_status_clean_interval` to `0` which will disable the timer (which I've just noticed isn't documented, so will update the docs now...)

I considered plugins, but sadly you can't set up a plugin in a vuex module (https://github.com/vuejs/vuex/issues/467) though this is apparently coming in Vue 3.x

**Comment by mrichar1** [comment]
OK, so I had an inspiration and have refactored things to do it 'properly'.

Expiring session ids now happens in `actionSequence`. This function is called from all the `actions` to get a `session id` - it now also uses setTimeout to arrange for that id to be deleted after the `action_status_clean_age` delay. Since this happens via an `action` it has access to `mutations` so it uses a `delete_status` mutation which means that `strict` mode now works.

This is pushed to `master` just now, and I'll release a new version soon.

**Comment by mrichar1** [comment]
Released in `0.1.1`
